mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,184 @@
+    where
+        for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<IoType>,
+    {
+        struct HandshakeFuture<'a, T>(&'a mut Context<T>);
+        impl<T> Future for HandshakeFuture<'_, T>
+        where
+            for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+        {
+            type Output = Result<()>;
+            fn poll(mut self: Pin<&mut Self>, ctx: &mut TaskContext) -> std::task::Poll<Self::Output> {
+                self.0
+                    .with_bio_async(ctx, |ssl_ctx| match ssl_ctx.handshake() {
+                        Err(Error::SslWantRead) | Err(Error::SslWantWrite) => Poll::Pending,
+                        Err(e) => Poll::Ready(Err(e)),
+                        Ok(()) => Poll::Ready(Ok(())),
+                    })
+                    .unwrap_or(Poll::Ready(Err(Error::NetSendFailed)))
+            }
+        }
+
+        self.prepare_handshake(io, hostname)?;
+
+        HandshakeFuture(self).await
+    }
+}
+
+impl<T: AsyncRead> AsyncRead for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &mut ReadBuf<'_>) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        self.with_bio_async(cx, |ssl_ctx| match ssl_ctx.recv(buf.initialize_unfilled()) {
+            Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(())),
+            Err(Error::SslWantRead) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+            Ok(i) => {
+                buf.advance(i);
+                Poll::Ready(Ok(()))
+            }
+        })
+        .unwrap_or_else(|| Poll::Ready(Err(crate::private::error_to_io_error(Error::NetRecvFailed))))
+    }
+}
+
+impl<T: AsyncWrite + Unpin> AsyncWrite for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_write(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &[u8]) -> Poll<IoResult<usize>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        let result = self
+            .with_bio_async(cx, |ssl_ctx| match ssl_ctx.async_write(buf) {
+                Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(0)),
+                Err(Error::SslWantWrite) => Poll::Pending,
+                Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+                Ok(i) => Poll::Ready(Ok(i)),
+            })
+            .unwrap_or_else(|| Poll::Ready(Err(crate::private::error_to_io_error(Error::NetSendFailed))));
+
+        cx.waker().clone().wake();
```

Comment:
This seems extremely inefficient: you're waking the task immediately every time this function is called (including if this returns Poll::Pending), basically resulting in continuous polling?